### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/TensorProduct): add `TensorProduct.map₂`

### PR DESCRIPTION
**Pull request by jjaassoonn**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by riccardobrasca** [comment]
Thanks!

bors d+

**Comment by eric-wieser** [comment]
Can you use this to golf the `Mul` instance on tensor products?

**Comment by jjaassoonn** [comment]
> Can you use this to golf the `Mul` instance on tensor products?

done

File: Mathlib/LinearAlgebra/TensorProduct.lean [hunk 1]
```diff
@@ -945,6 +945,10 @@ def homTensorHomMap : (M →ₗ[R] P) ⊗[R] (N →ₗ[R] Q) →ₗ[R] M ⊗[R]
 
 variable {R M N P Q}
 
+def map₂ (f : M →ₗ[R] P →ₗ[R] Q) (g : N →ₗ[R] S →ₗ[R] T) :
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/TensorProduct.lean:948]: Can you add a short docstring? Thanks!

**Comment by eric-wieser** [comment]
Once you've added the docstring, please run a benchmark (it will fail at the moment because the missing docstring will cause a linter error)

File: Mathlib/LinearAlgebra/TensorProduct.lean (hunk 1 above)
Comment (jjaassoonn) [reply Mathlib/LinearAlgebra/TensorProduct.lean:952]: Thanks for the reminder, didn't notice the missing doc string

**Comment by jjaassoonn** [comment]
!bench

File: Mathlib/LinearAlgebra/TensorProduct.lean [hunk 2]
```diff
@@ -945,6 +945,14 @@ def homTensorHomMap : (M →ₗ[R] P) ⊗[R] (N →ₗ[R] Q) →ₗ[R] M ⊗[R]
 
 variable {R M N P Q}
 
+/--
+Given any linear maps `f : M → Hom(P, Q)` and `g : N → Hom(S, T)`, there is a linear map
+`M ⊗ N → Hom(P ⊗ S, Q ⊗ T)` given by `m ⊗ n ↦ f m ⊗ g n`
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/TensorProduct.lean:950]: I think an explanation in terms of binary functions might be useful too, to explain the relation with `map`; that `map₂ f g (m1 ⊗ n1) (m2 ⊗ n2) = f m1 m2 ⊗ g m1 m2`.

Comment (jjaassoonn) [reply Mathlib/LinearAlgebra/TensorProduct.lean]: Thanks! I kept both version

**Comment by jjaassoonn** [comment]
How do we feel about the benchmark result. Is the slow down statistically insignificant?

**Comment by jjaassoonn** [comment]
bors merge

**Comment by eric-wieser** [comment]
I think it's ok; I tried golfing this a long time ago in #6195 and got a timeout; 5% is a much less bad outcome!

File: Mathlib/RingTheory/TensorProduct.lean [hunk 3]
```diff
@@ -213,40 +213,15 @@ variable [CommSemiring R]
...
 variable [NonUnitalNonAssocSemiring B] [Module R B] [SMulCommClass R B B] [IsScalarTower R B B]
 
-/-- (Implementation detail)
-The multiplication map on `A ⊗[R] B`,
-for a fixed pure tensor in the first argument,
-as an `R`-linear map.
--/
-def mulAux (a₁ : A) (b₁ : B) : A ⊗[R] B →ₗ[R] A ⊗[R] B :=
-  TensorProduct.map (LinearMap.mulLeft R a₁) (LinearMap.mulLeft R b₁)
-#align algebra.tensor_product.mul_aux Algebra.TensorProduct.mulAux
```
Comment (eric-wieser) [inline Mathlib/RingTheory/TensorProduct.lean:223]: These should be `noalign`ed

bors r-

bors d+

Comment (jjaassoonn) [reply Mathlib/RingTheory/TensorProduct.lean:223]: do you mean I should keep these code but just `noalign` them? Note that I have added 
```
#noalign algebra.tensor_product.mul_aux
#noalign algebra.tensor_product.mul_aux_apply
```

**Review by eric-wieser — COMMENTED** [review]
bors merge

**Pull request closed without merging**